src/engine/structures/lock.rs

```diff
@@ -0,0 +1,970 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::{
+    any::type_name,
+    collections::{HashMap, HashSet, VecDeque},
+    fmt::{self, Debug, Display},
+    future::Future,
+    ops::{Deref, DerefMut},
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, Mutex, MutexGuard,
+    },
+    task::{Context, Poll, Waker},
+};
+
+use futures::executor::block_on;
+use tokio::sync::{OwnedRwLockReadGuard, OwnedRwLockWriteGuard, RwLock};
+
+use crate::engine::{
+    structures::Table,
+    types::{AsUuid, LockKey, Name},
+};
+
+pub struct SharedGuard<G>(G);
+
+impl<T, G> Deref for SharedGuard<G>
+where
+    G: Deref<Target = T>,
+    T: ?Sized,
+{
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &*self.0
+    }
+}
+
+impl<G> Drop for SharedGuard<G> {
+    fn drop(&mut self) {
+        trace!("Dropping shared lock {}", type_name::<G>());
```

Comment:
Plz insert "on" into for greater consistency with other similar messages, as "Dropping shared lock on {}".